# <span style="color:red"> AI Booster Week 2 Evaluation </span>

### Complete this notebook according to the following requirements :

1. Fill-in the blanks by answering the questions
2. Comment clearly on each of your answer
3. Make sure your notebook is functioning as we will run it from beginning to end and each error will be penalized

-------------------------------------------------------------------------------------------

In [1]:
import pandas as pd
import numpy as np
import zipfile
import os

## import and unzip data files

Download the datasets from : https://github.com/jaotombo/data_science_from_scratch/tree/main/data

### I. Load data into pandas dataframe

Here, use the **read_csv method** to load data into pandas dataframe. 

Look how it is similar to what we did when using methods for the DataFrame class.

**Note:** Have a look at the pandas doc to understand the read_csv syntax

#### Start by readind the `victoria.csv` dataset and call it `df`

In [2]:
os.getcwd()

'C:\\Users\\yin\\Downloads'

In [3]:
## Your code here ##
# 
DATA_PATH = r'D:\database'
os.chdir (DATA_PATH)
# load data into a pandas dataframe named df.
df = pd.read_csv ('victoria.csv',sep = '*') 

In [4]:
# make a copy of dataframe which will be used after
df_copy = df.copy()

#### Check the type of the df object created 

In [5]:
## Your code here ##
type(df)

pandas.core.frame.DataFrame

In [6]:
# shift to see properties and methods
# or execute dir(df)
dir(df)

['T',
 '_AXIS_LEN',
 '_AXIS_ORDERS',
 '_AXIS_TO_AXIS_NUMBER',
 '_HANDLED_TYPES',
 '__abs__',
 '__add__',
 '__and__',
 '__annotations__',
 '__array__',
 '__array_priority__',
 '__array_ufunc__',
 '__array_wrap__',
 '__bool__',
 '__class__',
 '__contains__',
 '__copy__',
 '__deepcopy__',
 '__delattr__',
 '__delitem__',
 '__dict__',
 '__dir__',
 '__divmod__',
 '__doc__',
 '__eq__',
 '__finalize__',
 '__floordiv__',
 '__format__',
 '__ge__',
 '__getattr__',
 '__getattribute__',
 '__getitem__',
 '__getstate__',
 '__gt__',
 '__hash__',
 '__iadd__',
 '__iand__',
 '__ifloordiv__',
 '__imod__',
 '__imul__',
 '__init__',
 '__init_subclass__',
 '__invert__',
 '__ior__',
 '__ipow__',
 '__isub__',
 '__iter__',
 '__itruediv__',
 '__ixor__',
 '__le__',
 '__len__',
 '__lt__',
 '__matmul__',
 '__mod__',
 '__module__',
 '__mul__',
 '__ne__',
 '__neg__',
 '__new__',
 '__nonzero__',
 '__or__',
 '__pos__',
 '__pow__',
 '__radd__',
 '__rand__',
 '__rdivmod__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 

In [7]:
# shift-tab 1-4 times to show docs
#df()

#### Compute the dimensions of the dataframe (number of rows and number of columns)

In [8]:
## Your code here ##
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 453386 entries, 0 to 453385
Data columns (total 14 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   product_name      453386 non-null  object 
 1   mrp               453386 non-null  object 
 2   price             453386 non-null  object 
 3   pdp_url           453386 non-null  object 
 4   brand_name        453386 non-null  object 
 5   product_category  453386 non-null  object 
 6   retailer          453386 non-null  object 
 7   description       453386 non-null  object 
 8   rating            137734 non-null  float64
 9   review_count      137734 non-null  float64
 10  style_attributes  0 non-null       float64
 11  total_sizes       453386 non-null  object 
 12  available_size    453386 non-null  object 
 13  color             453386 non-null  object 
dtypes: float64(3), object(11)
memory usage: 48.4+ MB


In [9]:
#number of rows
len(df)

453386

In [10]:
#number of columns
len(df.columns)

14

In [11]:
#number of rows and columns
df.shape

(453386, 14)

#### Check the different columns of the df (type, length, etc.)
See how similar it is compared to the class approach

In [12]:
## Your code here ##
for column_name in df.columns :
    column_content = df[column_name]
    print(column_name)
    print(column_content)
    print()

product_name
0         Very Sexy Strappy Lace Thong Panty
1         Very Sexy Strappy Lace Thong Panty
2         Very Sexy Strappy Lace Thong Panty
3         Very Sexy Strappy Lace Thong Panty
4         Very Sexy Strappy Lace Thong Panty
                         ...                
453381      Dream Angels NEW! Lace Mini Slip
453382      Dream Angels NEW! Lace Mini Slip
453383      Dream Angels NEW! Lace Mini Slip
453384      Dream Angels NEW! Lace Mini Slip
453385      Dream Angels NEW! Lace Mini Slip
Name: product_name, Length: 453386, dtype: object

mrp
0         $14.50 
1         $14.50 
2         $14.50 
3         $14.50 
4         $14.50 
           ...   
453381    $48.00 
453382    $48.00 
453383    $48.00 
453384    $48.00 
453385    $48.00 
Name: mrp, Length: 453386, dtype: object

price
0         $14.50 
1         $14.50 
2         $14.50 
3         $14.50 
4         $14.50 
           ...   
453381    $48.00 
453382    $48.00 
453383    $48.00 
453384    $48.00 
453385    $

---
Use the method iloc (see the pandas doc) on df. What do you have as output? 

In [13]:
## Your code here ##
df.iloc[:,:]
# display the full dataframe(from first row to the end, and the first column to the last one.)

product_name      mrp    price  \
0       Very Sexy Strappy Lace Thong Panty  $14.50   $14.50    
1       Very Sexy Strappy Lace Thong Panty  $14.50   $14.50    
2       Very Sexy Strappy Lace Thong Panty  $14.50   $14.50    
3       Very Sexy Strappy Lace Thong Panty  $14.50   $14.50    
4       Very Sexy Strappy Lace Thong Panty  $14.50   $14.50    
...                                    ...      ...      ...   
453381    Dream Angels NEW! Lace Mini Slip  $48.00   $48.00    
453382    Dream Angels NEW! Lace Mini Slip  $48.00   $48.00    
453383    Dream Angels NEW! Lace Mini Slip  $48.00   $48.00    
453384    Dream Angels NEW! Lace Mini Slip  $48.00   $48.00    
453385    Dream Angels NEW! Lace Mini Slip  $48.00   $48.00    

                                                  pdp_url         brand_name  \
0       https://www.victoriassecret.com/panties/shop-a...  Victoria's Secret   
1       https://www.victoriassecret.com/panties/shop-a...  Victoria's Secret   
2       https://www.victoriassecret.com/panties/shop-a...  Victoria's Secret   
3       https://www.victoriassecret.com/panties/shop-a...  Victoria's Secret   
4       https://www.victoriassecret.com/panties/shop-a...  Victoria's Secret   
...                                                   ...                ...   
453381  https://www.victoriassecret.com/lingerie/shop-...  Victoria's Secret   
453382  https://www.victoriassecret.com/lingerie/shop-...  Victoria's Secret   
453383  https://www.victoriassecret.com/lingerie/shop-...  Victoria's Secret   
453384  https://www.victoriassecret.com/lingerie/shop-...  Victoria's Secret   
453385  https://www.victoriassecret.com/lingerie/shop-...  Victoria's Secret   

                product_category            retailer  \
0       Strappy Lace Thong Panty  Victoriassecret US   
1       Strappy Lace Thong Panty  Victoriassecret US   
2       Strappy Lace Thong Panty  Victoriassecret US   
3       Strappy Lace Thong Panty  Victoriassecret US   
4       Strappy Lace Thong Panty  Victoriassecret US   
...                          ...                 ...   
453381            Lace Mini Slip  Victoriassecret US   
453382            Lace Mini Slip  Victoriassecret US   
453383            Lace Mini Slip  Victoriassecret US   
453384            Lace Mini Slip  Victoriassecret US   
453385            Lace Mini Slip  Victoriassecret US   

                                              description  rating  \
0       Lots of cheek peek, pretty lace, a strappy bac...     NaN   
1       Lots of cheek peek, pretty lace, a strappy bac...     NaN   
2       Lots of cheek peek, pretty lace, a strappy bac...     NaN   
3       Lots of cheek peek, pretty lace, a strappy bac...     NaN   
4       Lots of cheek peek, pretty lace, a strappy bac...     NaN   
...                                                   ...     ...   
453381  All you need is a little lace, here in the cut...     NaN   
453382  All you need is a little lace, here in the cut...     NaN   
453383  All you need is a little lace, here in the cut...     NaN   
453384  All you need is a little lace, here in the cut...     NaN   
453385  All you need is a little lace, here in the cut...     NaN   

        review_count  style_attributes                  total_sizes  \
0                NaN               NaN  ["XS", "S", "M", "L", "XL"]   
1                NaN               NaN  ["XS", "S", "M", "L", "XL"]   
2                NaN               NaN  ["XS", "S", "M", "L", "XL"]   
3                NaN               NaN  ["XS", "S", "M", "L", "XL"]   
4                NaN               NaN  ["XS", "S", "M", "L", "XL"]   
...              ...               ...                          ...   
453381           NaN               NaN  ["XS", "S", "M", "L", "XL"]   
453382           NaN               NaN  ["XS", "S", "M", "L", "XL"]   
453383           NaN               NaN  ["XS", "S", "M", "L", "XL"]   
453384           NaN               NaN  ["XS", "S", "M", "L", "XL"]   
453385           

---
Which methods can be used to extract only the few first and last lines of the csv file? 

In [14]:
## Your code here ##
# head method can show first n rows of dataframe(default 5 rows)
df_head = df.head()
df_head

product_name      mrp    price  \
0  Very Sexy Strappy Lace Thong Panty  $14.50   $14.50    
1  Very Sexy Strappy Lace Thong Panty  $14.50   $14.50    
2  Very Sexy Strappy Lace Thong Panty  $14.50   $14.50    
3  Very Sexy Strappy Lace Thong Panty  $14.50   $14.50    
4  Very Sexy Strappy Lace Thong Panty  $14.50   $14.50    

                                             pdp_url         brand_name  \
0  https://www.victoriassecret.com/panties/shop-a...  Victoria's Secret   
1  https://www.victoriassecret.com/panties/shop-a...  Victoria's Secret   
2  https://www.victoriassecret.com/panties/shop-a...  Victoria's Secret   
3  https://www.victoriassecret.com/panties/shop-a...  Victoria's Secret   
4  https://www.victoriassecret.com/panties/shop-a...  Victoria's Secret   

           product_category            retailer  \
0  Strappy Lace Thong Panty  Victoriassecret US   
1  Strappy Lace Thong Panty  Victoriassecret US   
2  Strappy Lace Thong Panty  Victoriassecret US   
3  Strappy Lace Thong Panty  Victoriassecret US   
4  Strappy Lace Thong Panty  Victoriassecret US   

                                         description  rating  review_count  \
0  Lots of cheek peek, pretty lace, a strappy bac...     NaN           NaN   
1  Lots of cheek peek, pretty lace, a strappy bac...     NaN           NaN   
2  Lots of cheek peek, pretty lace, a strappy bac...     NaN           NaN   
3  Lots of cheek peek, pretty lace, a strappy bac...     NaN           NaN   
4  Lots of cheek peek, pretty lace, a strappy bac...     NaN           NaN   

   style_attributes                  total_sizes available_size        color  
0               NaN  ["XS", "S", "M", "L", "XL"]              S  peach melba  
1               NaN  ["XS", "S", "M", "L", "XL"]              S        black  
2               NaN  ["XS", "S", "M", "L", "XL"]              S    plum dust  
3               NaN  ["XS", "S", "M", "L", "XL"]              S  ensign blue  
4               NaN  ["XS", "S", "M", "L", "XL"]              S  fair orchid

In [15]:
df_tail = df.tail()
# tail method can show last n rows of dataframe(default 5 rows)
df_tail

product_name      mrp    price  \
453381  Dream Angels NEW! Lace Mini Slip  $48.00   $48.00    
453382  Dream Angels NEW! Lace Mini Slip  $48.00   $48.00    
453383  Dream Angels NEW! Lace Mini Slip  $48.00   $48.00    
453384  Dream Angels NEW! Lace Mini Slip  $48.00   $48.00    
453385  Dream Angels NEW! Lace Mini Slip  $48.00   $48.00    

                                                  pdp_url         brand_name  \
453381  https://www.victoriassecret.com/lingerie/shop-...  Victoria's Secret   
453382  https://www.victoriassecret.com/lingerie/shop-...  Victoria's Secret   
453383  https://www.victoriassecret.com/lingerie/shop-...  Victoria's Secret   
453384  https://www.victoriassecret.com/lingerie/shop-...  Victoria's Secret   
453385  https://www.victoriassecret.com/lingerie/shop-...  Victoria's Secret   

       product_category            retailer  \
453381   Lace Mini Slip  Victoriassecret US   
453382   Lace Mini Slip  Victoriassecret US   
453383   Lace Mini Slip  Victoriassecret US   
453384   Lace Mini Slip  Victoriassecret US   
453385   Lace Mini Slip  Victoriassecret US   

                                              description  rating  \
453381  All you need is a little lace, here in the cut...     NaN   
453382  All you need is a little lace, here in the cut...     NaN   
453383  All you need is a little lace, here in the cut...     NaN   
453384  All you need is a little lace, here in the cut...     NaN   
453385  All you need is a little lace, here in the cut...     NaN   

        review_count  style_attributes                  total_sizes  \
453381           NaN               NaN  ["XS", "S", "M", "L", "XL"]   
453382           NaN               NaN  ["XS", "S", "M", "L", "XL"]   
453383           NaN               NaN  ["XS", "S", "M", "L", "XL"]   
453384           NaN               NaN  ["XS", "S", "M", "L", "XL"]   
453385           NaN               NaN  ["XS", "S", "M", "L", "XL"]   

       available_size          color  
453381              L  coconut white  
453382              L   mulled grape  
453383              M     sheer pink  
453384              M  coconut white  
453385              M   mulled grape

---
Try to extract data from the 'product_name' column. 

Remember how we used it when defining __getitem__ in the dataFrame class.

In [16]:
## Your code here ##
df_pn = df.loc[:,"product_name"] #extract 'product_name' column and save it to df_pn
df_pn

0         Very Sexy Strappy Lace Thong Panty
1         Very Sexy Strappy Lace Thong Panty
2         Very Sexy Strappy Lace Thong Panty
3         Very Sexy Strappy Lace Thong Panty
4         Very Sexy Strappy Lace Thong Panty
                         ...                
453381      Dream Angels NEW! Lace Mini Slip
453382      Dream Angels NEW! Lace Mini Slip
453383      Dream Angels NEW! Lace Mini Slip
453384      Dream Angels NEW! Lace Mini Slip
453385      Dream Angels NEW! Lace Mini Slip
Name: product_name, Length: 453386, dtype: object

----
Use the describe method on df. What kind of output do you get? 

In [17]:
## Your code here ##
df.describe()
#we can get statistical data of the dataframe using describe method

rating  review_count  style_attributes
count  137734.000000  1.377340e+05               0.0
mean        4.165819  9.202884e+35               NaN
std         0.487524  1.162929e+37               NaN
min         0.000000  2.000000e+00               NaN
25%         4.000000  3.900000e+01               NaN
50%         4.300000  1.470000e+02               NaN
75%         4.500000  4.100000e+02               NaN
max         5.000000  1.560000e+38               NaN

-----
With Pandas, there are also some interesting methods that allow us to identify where you have null values (nan or empty).

Apply isnull (or isna) and sum methods on df. 

What do you get? 

In [18]:
## Your code here ##
df.isnull() #Detect missing values


product_name    mrp  price  pdp_url  brand_name  product_category  \
0              False  False  False    False       False             False   
1              False  False  False    False       False             False   
2              False  False  False    False       False             False   
3              False  False  False    False       False             False   
4              False  False  False    False       False             False   
...              ...    ...    ...      ...         ...               ...   
453381         False  False  False    False       False             False   
453382         False  False  False    False       False             False   
453383         False  False  False    False       False             False   
453384         False  False  False    False       False             False   
453385         False  False  False    False       False             False   

        retailer  description  rating  review_count  style_attributes  \
0          False        False    True          True              True   
1          False        False    True          True              True   
2          False        False    True          True              True   
3          False        False    True          True              True   
4          False        False    True          True              True   
...          ...          ...     ...           ...               ...   
453381     False        False    True          True              True   
453382     False        False    True          True              True   
453383     False        False    True          True              True   
453384     False        False    True          True              True   
453385     False        False    True          True              True   

        total_sizes  available_size  color  
0             False           False  False  
1             False           False  False  
2             False           False  False  
3             False           False  False  
4             False           False  False  
...             ...             ...    ...  
453381        False           False  False  
453382        False           False  False  
453383        False           False  False  
453384        False           False  False  
453385        False           False  False  

[453386 rows x 14 columns]

In [19]:
df.isnull().sum()
# count the missing value in each column

product_name             0
mrp                      0
price                    0
pdp_url                  0
brand_name               0
product_category         0
retailer                 0
description              0
rating              315652
review_count        315652
style_attributes    453386
total_sizes              0
available_size           0
color                    0
dtype: int64

---
Apply the value_counts() method to df considering the 'price' and 'product_name' columns

In [20]:
## Your code here ##
df['price'].value_counts()
#count the repetition of each value in 'price' column

$10.50     45832
$36.50     33365
$34.50     24636
$34.95     24148
$19.99     22784
           ...  
$16.95        24
$32.99        12
$64.95        12
$59.95        12
$9.95         11
Name: price, Length: 96, dtype: int64

In [21]:
## Your code here ##
df['product_name'].value_counts()
#count the repetition of each value in 'product_name' column 

Body by Victoria Perfect Coverage Bra              17922
Victoria Sport Incredible by Victoria Sport Bra    12274
Dream Angels Push-Up Bra                           11878
Very Sexy Push-Up Bra                              10985
Body by Victoria Demi Bra                          10881
                                                   ...  
Very Sexy XO Lace Bodysuit                             5
Very Sexy Embroidered Triangle Bra                     4
PINK Fresh & Clean Gym Kit                             3
PINK Coconut Waters Travel Kit                         3
Very Sexy Heart Lace Garter Belt                       2
Name: product_name, Length: 617, dtype: int64

----------

### II. Preprocess data

We are going to preprocess the dataset like we did in week 2 AI Booster, but using the Pandas methods instead. 

#### Step 1:

Convert `price` and `mrp` columns into quantitative (numerical)

First check all the values of 'mrp' and 'price' columns (by counting them for example).
And save them into new variables. 

In [22]:
# check the values of 'price' column by value_counts method
df['price'].value_counts()

$10.50     45832
$36.50     33365
$34.50     24636
$34.95     24148
$19.99     22784
           ...  
$16.95        24
$32.99        12
$64.95        12
$59.95        12
$9.95         11
Name: price, Length: 96, dtype: int64

In [23]:
# check the values of 'mrp' column
df['mrp'].value_counts()

$10.50     48627
$36.50     45968
$34.50     35026
$34.95     26374
$54.50     23152
           ...  
$24.00        29
$12.95        29
$16.95        24
$64.95        12
$9.95         11
Name: mrp, Length: 78, dtype: int64

In [24]:
## Your code here ##
df_price = df['price'].str.strip("$") #strip '$' in the 'price' column
df_mrp = df['mrp'].str.strip("$")     #strip '$' in the 'mrp' column

In [25]:
# just for test
print(df_price)
print(df_mrp)

0         14.50 
1         14.50 
2         14.50 
3         14.50 
4         14.50 
           ...  
453381    48.00 
453382    48.00 
453383    48.00 
453384    48.00 
453385    48.00 
Name: price, Length: 453386, dtype: object
0         14.50 
1         14.50 
2         14.50 
3         14.50 
4         14.50 
           ...  
453381    48.00 
453382    48.00 
453383    48.00 
453384    48.00 
453385    48.00 
Name: mrp, Length: 453386, dtype: object


Check the last values in the each of these new variables.

In [26]:
## Your code here ##
# using iloc method check the last values
df_price.iloc[-1],df_mrp.iloc[-1]

('48.00 ', '48.00 ')

We can see that the currency is indicated at the beginning of each value. 

How many types of different currencies do we have ?

In [27]:
## Your code here ##
# create a new series of the first character of each value in 'mrp' column
col_mrp = df['mrp'].str.slice(stop = 1)
# count the elements in new series
col_mrp.value_counts()

$    452880
R       350
₡       156
Name: mrp, dtype: int64

In [33]:
col_price = pd.DataFrame(price.strip()[0] for price in df['price'])
col_price.value_counts()

$    452880
R       350
₡       156
dtype: int64

---
Now create a "clean_price" function which takes a price as argument. 

Then if this element contains $, remove the dollar (put it into a float format).
<br>Otherwise do nothing. 

In [120]:
def clean_price(price):
    if '$' in price :
        price= price.strip("$") 
        return float(price)
    else: 
        price = np.nan
   

Apply this function to 'mrp' and 'price' columns, using apply. 

Here the **apply method** is only valid for a specific column of a dataset. 
Use **applymap** if you need to apply a function to 2 or more columns.

In [121]:
# Test function
df_mrp_clean = df['mrp'].apply(clean_price)

In [122]:
df_price_clean = df['price'].apply(clean_price)

In [123]:
df_mrp_clean,df_price_clean

(0         14.5
 1         14.5
 2         14.5
 3         14.5
 4         14.5
           ... 
 453381    48.0
 453382    48.0
 453383    48.0
 453384    48.0
 453385    48.0
 Name: mrp, Length: 453386, dtype: float64,
 0         14.5
 1         14.5
 2         14.5
 3         14.5
 4         14.5
           ... 
 453381    48.0
 453382    48.0
 453383    48.0
 453384    48.0
 453385    48.0
 Name: price, Length: 453386, dtype: float64)

In [124]:
# Apply the function to two columns by applymap
df[['price','mrp']].applymap(clean_price)

price   mrp
0        14.5  14.5
1        14.5  14.5
2        14.5  14.5
3        14.5  14.5
4        14.5  14.5
...       ...   ...
453381   48.0  48.0
453382   48.0  48.0
453383   48.0  48.0
453384   48.0  48.0
453385   48.0  48.0

[453386 rows x 2 columns]

In [125]:
df.price = df_price_clean
df.mrp = df_mrp_clean

Describe your dataset (df) once more. Has anything changed? If so why ?

In [126]:
df.describe()

mrp          price         rating  review_count  \
count  452880.000000  452880.000000  137734.000000  1.377340e+05   
mean       36.028995      32.338643       4.165819  9.202884e+35   
std        14.762885      15.657394       0.487524  1.162929e+37   
min         8.500000       2.990000       0.000000  2.000000e+00   
25%        29.500000      19.990000       4.000000  3.900000e+01   
50%        36.000000      34.500000       4.300000  1.470000e+02   
75%        46.500000      42.500000       4.500000  4.100000e+02   
max        98.000000      98.000000       5.000000  1.560000e+38   

       style_attributes  
count               0.0  
mean                NaN  
std                 NaN  
min                 NaN  
25%                 NaN  
50%                 NaN  
75%                 NaN  
max                 NaN

#### <span style="color:red">Do you notice anything bizarre ? If so should you do something ?</span>

'mrp' and 'price' should count 453386. There are some missing values in 'mrp','price' columns. 
We can replace them by another value.

----

#### Step 2:

Fill missing values by columns' medians

In [127]:
#test with 'rating' column
#compute the median by median method
median = df['rating'][df['rating']!=''].median()
#fill missing values by median
df.loc[df['rating'] == '', 'rating'] = median

df['rating'].value_counts()

4.4    24787
4.5    20185
4.3    20007
4.0    12024
4.2     8484
4.6     7271
3.7     7044
3.9     6132
4.1     5158
3.8     4618
3.2     4001
3.6     3927
4.7     2929
4.8     2330
3.3     2315
5.0     1499
3.5     1276
1.0      859
4.9      807
3.0      423
2.9      350
2.6      324
3.1      312
3.4      193
2.8      175
0.0      112
2.0       60
2.7       54
2.5       48
2.3       18
2.4       12
Name: rating, dtype: int64

In [128]:
len(df.rating)

453386

Apply the **median() method** to df on the relevant features (columns)

And save the output into a variable called medians.

In [129]:
## Your code here ##
#DataFrame.median(axis=_NoDefault.no_default, skipna=True, level=None, numeric_only=None, **kwargs)
price_med = df['price'].median()
mrp_med = df['mrp'].median()
review_med = df['review_count'].median()
rating_med = df['rating'].median()

In [130]:
df.loc[df['price'] == '', 'price'] = price_med
df.loc[df['mrp'] == '', 'mrp'] = mrp_med
df.loc[df['review_count'] == '', 'review_count'] = review_med

In [131]:
len(df.price),len(df.mrp),len(df.review_count)

(453386, 453386, 453386)

----
The fillna method exists in Pandas. It does exactly what we saw in the class refactoring methods. 
(https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.fillna.html)

Apply this one to the 'mrp', 'price', 'rating' and 'review_count' columns in df, replacing nan values by medians.

Then again describe your dataframe. Anything different ?

In [132]:
## Your code here ##
df['rating'].fillna(rating_med, inplace = False)
df['review_count'].fillna(review_med, inplace = False)
df['mrp'].fillna(mrp_med, inplace = False)
df['price'].fillna(price_med, inplace = False)

0         14.5
1         14.5
2         14.5
3         14.5
4         14.5
          ... 
453381    48.0
453382    48.0
453383    48.0
453384    48.0
453385    48.0
Name: price, Length: 453386, dtype: float64

In [133]:
## Your code here ##
df.describe()
# the missing values are not be replaced

mrp          price         rating  review_count  \
count  452880.000000  452880.000000  137734.000000  1.377340e+05   
mean       36.028995      32.338643       4.165819  9.202884e+35   
std        14.762885      15.657394       0.487524  1.162929e+37   
min         8.500000       2.990000       0.000000  2.000000e+00   
25%        29.500000      19.990000       4.000000  3.900000e+01   
50%        36.000000      34.500000       4.300000  1.470000e+02   
75%        46.500000      42.500000       4.500000  4.100000e+02   
max        98.000000      98.000000       5.000000  1.560000e+38   

       style_attributes  
count               0.0  
mean                NaN  
std                 NaN  
min                 NaN  
25%                 NaN  
50%                 NaN  
75%                 NaN  
max                 NaN

Apply fillna to df using inplace = True <br>
What does is accomplish precisely ? <br> Provide an illustration. 


In [134]:
## Your code here ##
df['rating'].fillna(rating_med, inplace = True)
df['review_count'].fillna(review_med, inplace = True)
df['mrp'].fillna(mrp_med, inplace = True)
df['price'].fillna(price_med, inplace = True)

In [135]:
## Your code here ##
df.describe()
# the missing values are replaced

mrp          price         rating  review_count  \
count  453386.000000  453386.000000  453386.000000  4.533860e+05   
mean       36.028963      32.341055       4.259237  2.795741e+35   
std        14.754645      15.648821       0.275703  6.423671e+36   
min         8.500000       2.990000       0.000000  2.000000e+00   
25%        29.500000      19.990000       4.300000  1.470000e+02   
50%        36.000000      34.500000       4.300000  1.470000e+02   
75%        46.500000      42.500000       4.300000  1.470000e+02   
max        98.000000      98.000000       5.000000  1.560000e+38   

       style_attributes  
count               0.0  
mean                NaN  
std                 NaN  
min                 NaN  
25%                 NaN  
50%                 NaN  
75%                 NaN  
max                 NaN

In [136]:
df.isnull().sum()

product_name             0
mrp                      0
price                    0
pdp_url                  0
brand_name               0
product_category         0
retailer                 0
description              0
rating                   0
review_count             0
style_attributes    453386
total_sizes              0
available_size           0
color                    0
dtype: int64

For further use, let's make a preprocess function that cleans the price and imputes missing values <br>
Justify your imputation method (which imputation method did you chose and why ?)

In [ ]:
def preproc_try(df):
    for column in df:
        column = column.apply(clean_price)
    
            
        
        

In [137]:
## Your code here ##
# Define a preproc function to clean the dataframe
def preproc (df):
    # apply clean_price to price and mrp columns
    df['price'] = df['price'].apply(clean_price)
    df['mrp'] = df['mrp'].apply(clean_price)
    
    # compute the median of each column
    price_med = df['price'].median()
    mrp_med = df['mrp'].median()
    review_med = df['review_count'].median()
    rating_med = df['rating'].median()
    
    # using fillna method to impute missing values
    df['review_count'].fillna(review_med, inplace = True)
    df['rating'].fillna(rating_med, inplace = True)
    df['price'].fillna(price_med, inplace = True)
    df['mrp'].fillna(mrp_med, inplace = True)
    
    return df


In [138]:
## Your code here ##
# test the funciton
preproc(df_copy)

product_name   mrp  price  \
0       Very Sexy Strappy Lace Thong Panty  14.5   14.5   
1       Very Sexy Strappy Lace Thong Panty  14.5   14.5   
2       Very Sexy Strappy Lace Thong Panty  14.5   14.5   
3       Very Sexy Strappy Lace Thong Panty  14.5   14.5   
4       Very Sexy Strappy Lace Thong Panty  14.5   14.5   
...                                    ...   ...    ...   
453381    Dream Angels NEW! Lace Mini Slip  48.0   48.0   
453382    Dream Angels NEW! Lace Mini Slip  48.0   48.0   
453383    Dream Angels NEW! Lace Mini Slip  48.0   48.0   
453384    Dream Angels NEW! Lace Mini Slip  48.0   48.0   
453385    Dream Angels NEW! Lace Mini Slip  48.0   48.0   

                                                  pdp_url         brand_name  \
0       https://www.victoriassecret.com/panties/shop-a...  Victoria's Secret   
1       https://www.victoriassecret.com/panties/shop-a...  Victoria's Secret   
2       https://www.victoriassecret.com/panties/shop-a...  Victoria's Secret   
3       https://www.victoriassecret.com/panties/shop-a...  Victoria's Secret   
4       https://www.victoriassecret.com/panties/shop-a...  Victoria's Secret   
...                                                   ...                ...   
453381  https://www.victoriassecret.com/lingerie/shop-...  Victoria's Secret   
453382  https://www.victoriassecret.com/lingerie/shop-...  Victoria's Secret   
453383  https://www.victoriassecret.com/lingerie/shop-...  Victoria's Secret   
453384  https://www.victoriassecret.com/lingerie/shop-...  Victoria's Secret   
453385  https://www.victoriassecret.com/lingerie/shop-...  Victoria's Secret   

                product_category            retailer  \
0       Strappy Lace Thong Panty  Victoriassecret US   
1       Strappy Lace Thong Panty  Victoriassecret US   
2       Strappy Lace Thong Panty  Victoriassecret US   
3       Strappy Lace Thong Panty  Victoriassecret US   
4       Strappy Lace Thong Panty  Victoriassecret US   
...                          ...                 ...   
453381            Lace Mini Slip  Victoriassecret US   
453382            Lace Mini Slip  Victoriassecret US   
453383            Lace Mini Slip  Victoriassecret US   
453384            Lace Mini Slip  Victoriassecret US   
453385            Lace Mini Slip  Victoriassecret US   

                                              description  rating  \
0       Lots of cheek peek, pretty lace, a strappy bac...     4.3   
1       Lots of cheek peek, pretty lace, a strappy bac...     4.3   
2       Lots of cheek peek, pretty lace, a strappy bac...     4.3   
3       Lots of cheek peek, pretty lace, a strappy bac...     4.3   
4       Lots of cheek peek, pretty lace, a strappy bac...     4.3   
...                                                   ...     ...   
453381  All you need is a little lace, here in the cut...     4.3   
453382  All you need is a little lace, here in the cut...     4.3   
453383  All you need is a little lace, here in the cut...     4.3   
453384  All you need is a little lace, here in the cut...     4.3   
453385  All you need is a little lace, here in the cut...     4.3   

        review_count  style_attributes                  total_sizes  \
0              147.0               NaN  ["XS", "S", "M", "L", "XL"]   
1              147.0               NaN  ["XS", "S", "M", "L", "XL"]   
2              147.0               NaN  ["XS", "S", "M", "L", "XL"]   
3              147.0               NaN  ["XS", "S", "M", "L", "XL"]   
4              147.0               NaN  ["XS", "S", "M", "L", "XL"]   
...              ...               ...                          ...   
453381         147.0               NaN  ["XS", "S", "M", "L", "XL"]   
453382         147.0               NaN  ["XS", "S", "M", "L", "XL"]   
453383         147.0               NaN  ["XS", "S", "M", "L", "XL"]   
453384         147.0               NaN  ["XS", "S", "M", "L", "XL"]   
453385         147.0               NaN  ["XS", "S", "M", "L", "XL"]   

     

### III. Data lookup



Let's repeat our previous exploration, but this time using Pandas methods. 

#### Question 1

Which product are the most expensive? what are the indices of the correponding rows and correponding products?

In [139]:
#extract the most expensive products
#if the price equals to the maximumn price, then put the price into a new dataframe
df_price_max = df[df['price'] == df['price'].max()]
#display product name and price
df_price_max[['product_name','price']]

product_name  price
578         Very Sexy Fishnet Lace Kimono   98.0
579         Very Sexy Fishnet Lace Kimono   98.0
22386   Very Sexy NEW! Keyhole Lace Teddy   98.0
22387   Very Sexy NEW! Keyhole Lace Teddy   98.0
22388   Very Sexy NEW! Keyhole Lace Teddy   98.0
...                                   ...    ...
444135  Very Sexy NEW! Keyhole Lace Teddy   98.0
444136  Very Sexy NEW! Keyhole Lace Teddy   98.0
444137  Very Sexy NEW! Keyhole Lace Teddy   98.0
444138  Very Sexy NEW! Keyhole Lace Teddy   98.0
444139  Very Sexy NEW! Keyhole Lace Teddy   98.0

[143 rows x 2 columns]

#### Question 2

What are the cheapest products which exists in size 36B? In which colors are they available?

In [140]:
#extract the rows containg 36B with df.str.contains method.
#here can also use df['available_size']== '36B'
#but if there is a list of string like ['36B','35C','37B',...], contains method is better
df_36b = df[df['available_size'].str.contains('36B')]
#extract cheapest product in 36B
df_36b_min = df_36b[df_36b['price']==df_36b['price'].min()]
#and display the color
df_36b_min[['product_name','price','available_size','color']]

product_name  price  \
181575  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181615  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181616  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181617  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181642  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181643  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181670  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181671  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181867  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181868  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181869  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181870  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181871  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181872  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181873  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181874  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181875  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181876  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181877  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181878  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181879  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181880  Victoria Sport Lightweight by Victoria Sport Bra    5.0   
181881  Victoria Sport Lightweight by Victoria Sport Bra    5.0   

       available_size                      color  
181575            36B                 Black Marl  
181615            36B    Grey Oasis Monofilament  
181616            36B                Border Line  
181617            36B         Black Monofilament  
181642            36B           White Colorblock  
181643            36B         Inkblot Colorblock  
181670            36B       Black Blocked Curves  
181671            36B    Tinted Plum Mesh Curves  
181867            36B           Firework Texture  
181868            36B                 Seychelles  
181869            36B         Cool Rainbow Rings  
181870            36B             Endless Orange  
181871            36B                      White  
181872            36B                      Black  
181873            36B                 Grey Oasis  
181874            36B              Red Geometric  
181875            36B        Red Sacred Geometry  
181876            36B               Hello Lovely  
181877            36B        Cool Textile Stripe  
181878            36B                Wavy Arrows  
181879            36B             Purple Rapture  
181880            36B            Blue Camo Zebra  
181881            36B  Warm Spraypaint Triangles

#### Question 3

We can do the same, but checking for products for which `total_sizes` contains '36B'

In [141]:
#extract the rows containg 36B
df_36b_total = df[df['total_sizes'].str.contains('36B')]
df_36b_total

product_name    mrp  price  \
108                       Body by Victoria Demi Bra  54.50  19.99   
109                       Body by Victoria Demi Bra  54.50  19.99   
110                       Body by Victoria Demi Bra  54.50  19.99   
111                       Body by Victoria Demi Bra  54.50  19.99   
112                       Body by Victoria Demi Bra  54.50  19.99   
...                                             ...    ...    ...   
453036  PINK NEW! Wear Everywhere Lightly Lined Bra  36.95  36.95   
453037  PINK NEW! Wear Everywhere Lightly Lined Bra  36.95  36.95   
453038  PINK NEW! Wear Everywhere Lightly Lined Bra  36.95  36.95   
453039  PINK NEW! Wear Everywhere Lightly Lined Bra  36.95  36.95   
453040  PINK NEW! Wear Everywhere Lightly Lined Bra  36.95  36.95   

                                                  pdp_url  \
108     https://www.victoriassecret.com/bras/shop-all-...   
109     https://www.victoriassecret.com/bras/shop-all-...   
110     https://www.victoriassecret.com/bras/shop-all-...   
111     https://www.victoriassecret.com/bras/shop-all-...   
112     https://www.victoriassecret.com/bras/shop-all-...   
...                                                   ...   
453036  https://www.victoriassecret.com/pink/panties/w...   
453037  https://www.victoriassecret.com/pink/panties/w...   
453038  https://www.victoriassecret.com/pink/panties/w...   
453039  https://www.victoriassecret.com/pink/panties/w...   
453040  https://www.victoriassecret.com/pink/panties/w...   

                    brand_name                   product_category  \
108          Victoria's Secret                           Demi Bra   
109          Victoria's Secret                           Demi Bra   
110          Victoria's Secret                           Demi Bra   
111          Victoria's Secret                           Demi Bra   
112          Victoria's Secret                           Demi Bra   
...                        ...                                ...   
453036  Victoria's Secret Pink  Wear Everywhere Lightly Lined Bra   
453037  Victoria's Secret Pink  Wear Everywhere Lightly Lined Bra   
453038  Victoria's Secret Pink  Wear Everywhere Lightly Lined Bra   
453039  Victoria's Secret Pink  Wear Everywhere Lightly Lined Bra   
453040  Victoria's Secret Pink  Wear Everywhere Lightly Lined Bra   

                  retailer                                        description  \
108     Victoriassecret US  Sexy comfort and a sleek shape start with low-...   
109     Victoriassecret US  Sexy comfort and a sleek shape start with low-...   
110     Victoriassecret US  Sexy comfort and a sleek shape start with low-...   
111     Victoriassecret US  Sexy comfort and a sleek shape start with low-...   
112     Victoriassecret US  Sexy comfort and a sleek shape start with low-...   
...                    ...                                                ...   
453036  Victoriassecret US  A true wardrobe essential made new with even s...   
453037  Victoriassecret US  A true wardrobe essential made new with even s...   
453038  Victoriassecret US  A true wardrobe essential made new with even s...   
453039  Victoriassecret US  A true wardrobe essential made new with even s...   
453040  Victoriassecret US  A true wardrobe essential made new with even s...   

        rating  review_count  style_attributes  \
108        4.3         147.0               NaN   
109        4.3         147.0               NaN   
110        4.3         147.0               NaN   
111        4.3         147.0               NaN   
112        4.3         147.0               NaN   
...        ...           ...               ...   
453036     4.3         147.0               NaN   
453037     4.3         147.0               NaN   
453038     4.3         147.0               NaN   
453039     4.3         147.0               NaN   
453040     4.3         147.0               NaN   

                                              total_sizes available_size  

#### Question 4

What are the cheapest 'sport bras'?

In [142]:
# using the same method as above, but search in product_category
df_sport = df.loc[df['product_category'].str.contains('Sport Bra')]
df_sport_min = df_sport[df_sport['price']==df_sport['price'].min()]
df_sport_min

product_name   mrp  price  \
161993  Victoria Sport Strappy Triangle Sport Bra  20.0    5.0   
161994  Victoria Sport Strappy Triangle Sport Bra  20.0    5.0   
161995  Victoria Sport Strappy Triangle Sport Bra  20.0    5.0   
161996  Victoria Sport Strappy Triangle Sport Bra  20.0    5.0   
161997  Victoria Sport Strappy Triangle Sport Bra  20.0    5.0   
...                                           ...   ...    ...   
256450  Victoria Sport Strappy Triangle Sport Bra  20.0    5.0   
256451  Victoria Sport Strappy Triangle Sport Bra  20.0    5.0   
256452  Victoria Sport Strappy Triangle Sport Bra  20.0    5.0   
256453  Victoria Sport Strappy Triangle Sport Bra  20.0    5.0   
256454  Victoria Sport Strappy Triangle Sport Bra  20.0    5.0   

                                                  pdp_url         brand_name  \
161993  https://www.victoriassecret.com/bras/shop-all-...  Victoria's Secret   
161994  https://www.victoriassecret.com/bras/shop-all-...  Victoria's Secret   
161995  https://www.victoriassecret.com/bras/shop-all-...  Victoria's Secret   
161996  https://www.victoriassecret.com/bras/shop-all-...  Victoria's Secret   
161997  https://www.victoriassecret.com/bras/shop-all-...  Victoria's Secret   
...                                                   ...                ...   
256450  https://www.victoriassecret.com/victorias-secr...  Victoria's Secret   
256451  https://www.victoriassecret.com/victorias-secr...  Victoria's Secret   
256452  https://www.victoriassecret.com/victorias-secr...  Victoria's Secret   
256453  https://www.victoriassecret.com/victorias-secr...  Victoria's Secret   
256454  https://www.victoriassecret.com/victorias-secr...  Victoria's Secret   

                  product_category            retailer  \
161993  Strappy Triangle Sport Bra  Victoriassecret US   
161994  Strappy Triangle Sport Bra  Victoriassecret US   
161995  Strappy Triangle Sport Bra  Victoriassecret US   
161996  Strappy Triangle Sport Bra  Victoriassecret US   
161997  Strappy Triangle Sport Bra  Victoriassecret US   
...                            ...                 ...   
256450  Strappy Triangle Sport Bra  Victoriassecret US   
256451  Strappy Triangle Sport Bra  Victoriassecret US   
256452  Strappy Triangle Sport Bra  Victoriassecret US   
256453  Strappy Triangle Sport Bra  Victoriassecret US   
256454  Strappy Triangle Sport Bra  Victoriassecret US   

                                              description  rating  \
161993  A sporty twist on a classic shape: light suppo...     4.3   
161994  A sporty twist on a classic shape: light suppo...     4.3   
161995  A sporty twist on a classic shape: light suppo...     4.3   
161996  A sporty twist on a classic shape: light suppo...     4.3   
161997  A sporty twist on a classic shape: light suppo...     4.3   
...                                                   ...     ...   
256450  A sporty twist on a classic shape: light suppo...     4.3   
256451  A sporty twist on a classic shape: light suppo...     4.3   
256452  A sporty twist on a classic shape: light suppo...     4.3   
256453  A sporty twist on a classic shape: light suppo...     4.3   
256454  A sporty twist on a classic shape: light suppo...     4.3   

        review_count  style_attributes            total_sizes available_size  \
161993         147.0               NaN  ["XS", "S", "M", "L"]              S   
161994         147.0               NaN  ["XS", "S", "M", "L"]              S   
161995         147.0               NaN  ["XS", "S", "M", "L"]              S   
161996         147.0               NaN  ["XS", "S", "M", "L"]              S   
161997         147.0               NaN  ["XS", "S", "M", "L"]              S   
...              ...               ...                    ...            ...   
256450         147.0               NaN  ["XS", "S", "M", "L"]              M   
256451         147.0               NaN  ["XS", "S", "M", "L"]              M   
256452         147.0               Na

#### Question 5
What is the most expensive belt? in which color is it available?

In [143]:
df_belt = df.loc[df['product_category'].str.contains('Belt')]
df_belt_max = df_belt[df_belt['price']==df_belt['price'].max()]
df_belt_max

product_name   mrp  price  \
293828  Very Sexy Heart Lace Garter Belt  58.0  19.99   
293829  Very Sexy Heart Lace Garter Belt  58.0  19.99   

                                                  pdp_url         brand_name  \
293828  https://www.victoriassecret.com/panties/shop-a...  Victoria's Secret   
293829  https://www.victoriassecret.com/panties/shop-a...  Victoria's Secret   

              product_category            retailer  \
293828  Heart Lace Garter Belt  Victoriassecret US   
293829  Heart Lace Garter Belt  Victoriassecret US   

                                              description  rating  \
293828  Queen of hearts: this daring garter belt starr...     4.3   
293829  Queen of hearts: this daring garter belt starr...     4.3   

        review_count  style_attributes       total_sizes available_size  \
293828         147.0               NaN  ["XS/S", "M/L "]           XS/S   
293829         147.0               NaN  ["XS/S", "M/L "]            M/L   

                     color  
293828  black hearts print  
293829  black hearts print

#### Question 6

What is the most common color of the best rated products?

What is then the best rated product available in the most popular color?

In [144]:
# extract the products with the highest rating
df_rating_max = df[df['rating'] == df['rating'].max()]
# look for the most common color by value_counts method
# then, extract the color with the highest count number
df_commoncolor = df_rating_max['color'].value_counts().idxmax()
# using str.contains method to extract the products available in the most common color from the df_rating_max which contains the products with the highest rating
df_best_pro = df_rating_max.loc[df['color'].str.contains(df_commoncolor)]

df_best_pro

product_name   mrp  price  \
174508  Victoria Sport NEW! Lightweight by Victoria Sp...  29.5   29.5   
174512  Victoria Sport NEW! Lightweight by Victoria Sp...  29.5   29.5   
174516  Victoria Sport NEW! Lightweight by Victoria Sp...  29.5   29.5   
174518  Victoria Sport NEW! Lightweight by Victoria Sp...  29.5   29.5   
174522  Victoria Sport NEW! Lightweight by Victoria Sp...  29.5   29.5   
...                                                   ...   ...    ...   
403271  Victoria Sport NEW! Lightweight by Victoria Sp...  29.5   29.5   
403275  Victoria Sport NEW! Lightweight by Victoria Sp...  29.5   29.5   
403279  Victoria Sport NEW! Lightweight by Victoria Sp...  29.5   29.5   
403282  Victoria Sport NEW! Lightweight by Victoria Sp...  29.5   29.5   
403285  Victoria Sport NEW! Lightweight by Victoria Sp...  29.5   29.5   

                                                  pdp_url         brand_name  \
174508  https://www.victoriassecret.com/bras/shop-all-...  Victoria's Secret   
174512  https://www.victoriassecret.com/bras/shop-all-...  Victoria's Secret   
174516  https://www.victoriassecret.com/bras/shop-all-...  Victoria's Secret   
174518  https://www.victoriassecret.com/bras/shop-all-...  Victoria's Secret   
174522  https://www.victoriassecret.com/bras/shop-all-...  Victoria's Secret   
...                                                   ...                ...   
403271  https://www.victoriassecret.com/victorias-secr...  Victoria's Secret   
403275  https://www.victoriassecret.com/victorias-secr...  Victoria's Secret   
403279  https://www.victoriassecret.com/victorias-secr...  Victoria's Secret   
403282  https://www.victoriassecret.com/victorias-secr...  Victoria's Secret   
403285  https://www.victoriassecret.com/victorias-secr...  Victoria's Secret   

                                       product_category            retailer  \
174508  Lightweight by Victoria Sport Strappy Sport Bra  Victoriassecret US   
174512  Lightweight by Victoria Sport Strappy Sport Bra  Victoriassecret US   
174516  Lightweight by Victoria Sport Strappy Sport Bra  Victoriassecret US   
174518  Lightweight by Victoria Sport Strappy Sport Bra  Victoriassecret US   
174522  Lightweight by Victoria Sport Strappy Sport Bra  Victoriassecret US   
...                                                 ...                 ...   
403271  Lightweight by Victoria Sport Strappy Sport Bra  Victoriassecret US   
403275  Lightweight by Victoria Sport Strappy Sport Bra  Victoriassecret US   
403279  Lightweight by Victoria Sport Strappy Sport Bra  Victoriassecret US   
403282  Lightweight by Victoria Sport Strappy Sport Bra  Victoriassecret US   
403285  Lightweight by Victoria Sport Strappy Sport Bra  Victoriassecret US   

                                              description  rating  \
174508  From cycling to circuit training, this medium-...     5.0   
174512  From cycling to circuit training, this medium-...     5.0   
174516  From cycling to circuit training, this medium-...     5.0   
174518  From cycling to circuit training, this medium-...     5.0   
174522  From cycling to circuit training, this medium-...     5.0   
...                                                   ...     ...   
403271  Our #1 sport bra for a reason: with a supersof...     5.0   
403275  Our #1 sport bra for a reason: with a supersof...     5.0   
403279  Our #1 sport bra for a reason: with a supersof...     5.0   
403282  Our #1 sport bra for a reason: with a supersof...     5.0   
403285  Our #1 sport bra for a reason: with a supersof...     5.0   

        review_count  style_attributes  \
174508           3.0               NaN   
174512           3.0               NaN   
174516           3.0               NaN   
174518           3.0               NaN   
174522           3.0               NaN   
...              ...               ...   
403271           3.0               NaN   
403275           3.0               NaN   
403279           3.0               NaN 

---

---